contracts/Unipool.sol

```diff
@@ -62,8 +72,25 @@ contract Unipool is LPTokenWrapper {
         _;
     }
 
-    constructor(IERC20 _uniswapTokenExchange) public {
+    constructor(IERC20 _uniswapTokenExchange, IUniswapV2Router01 _uniswapRouter, IERC20 _tradedToken) public {
         uniswapTokenExchange = _uniswapTokenExchange;
+        uniswapRouter = _uniswapRouter;
+        tradedToken = _tradedToken;
+
+        IUniswapV2Pair pair = IUniswapV2Pair(address(uniswapTokenExchange));
+
+        if (pair.token0() != address(tradedToken) && pair.token1() != address(tradedToken)) {
+            // We can only reinvest if we're giving rewards in one side of the pair (currently)
+            reinvestableToken = IERC20(address(0));
+        } else {
+            if (pair.token0() == address(tradedToken)) {
+                reinvestableToken = IERC20(pair.token1());
+            } else {
+                reinvestableToken = IERC20(pair.token0());
+            }
```

Comment:
Doing it that way uses quite a bit more gas for every request, right? The reason I cached `reinvestableToken` on construction, and do all these checks up front, is because these values shouldn't ever change. We have to ask the pair which token is which every time if we do it that way.

I can see how it's clearer, but it's a performance concern. It also doesn't really mitigate any attacks that I can see; if someone is engineering an evil pool contract that is going to shuffle around its responses for `token0()` and `token1()`, they won't be stopped by us checking every time. The defense against that is not to use farm contracts targeting weird pools.

Reply:
I did break this out into its own method though. It's definitely complicated enough to warrant it.

Reply:
I typically favour clarity over performance because security is so important, especially considering the environments that 1Hive apps live on, namely xDai and in future roll-ups, where the extra gas cost creates negligible, if any, extra cost to the user. However, if both can be achieved obviously that's ideal. 

I'll accept the above but `reinvestableToken` should be renamed because it's kind of misleading since both tokens are going to effectively be reinvested. Perhaps `pairedToken` or something that indicates it's the other side of the pair to the reward token. 

Also you can remove one of those pair.token() requests if you re-arrange those if statements. 

Also I would have the interface return the IERC20 interface that this contract references, then there's no need to do `address(contract)` everywhere.

Reply:
Can't get away from `address()` casts everywhere, but I did clean some up. I also added a modifier method that just checks whether the relevant address is null or not - sort of the best of both worlds.